File: tui/src/scala/tui/Constraint.scala
package tui

//#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
sealed trait Constraint {
  def apply(length: Int): Int =
    this match {
      case Constraint.Percentage(p)   => length * p / 100
      case Constraint.Ratio(num, den) => num * length / den
      case Constraint.Length(l)       => length.min(l)
      case Constraint.Max(m)          => length.min(m)
      case Constraint.Min(m)          => length.max(m)
    }
}

object Constraint {
  case class Percentage(p: Int) extends Constraint {
    require(p >= 0 && p <= 100)
  }
  case class Ratio(num: Int, den: Int) extends Constraint
  case class Length(l: Int) extends Constraint
  case class Max(m: Int) extends Constraint
  case class Min(m: Int) extends Constraint
}

File: tui/src/java/tui/Constraint.java
package tui;

public sealed interface Constraint
    permits Constraint.Percentage,
        Constraint.Ratio,
        Constraint.Length,
        Constraint.Max,
        Constraint.Min {

  default int apply(int length) {
    return switch (this) {
      case Percentage p -> length * p.p() / 100;
      case Ratio r -> r.num() * length / r.den();
      case Length l -> Math.min(length, l.l());
      case Max m -> Math.min(length, m.m());
      case Min m -> Math.max(length, m.m());
    };
  }

  record Percentage(int p) implements Constraint {
    public Percentage {
      if (p < 0 || p > 100) {
        throw new IllegalArgumentException("Percentage out of range: " + p);
      }
    }
  }

  record Ratio(int num, int den) implements Constraint {}

  record Length(int l) implements Constraint {}

  record Max(int m) implements Constraint {}

  record Min(int m) implements Constraint {}
}
